### ISSUE TITLE
AllowAnonymous being ignored with multiple auth schemes

### ISSUE THREAD

**Issue by pbakun** [dotnet/aspnetcore#21800]
I have an issue with Authorization for my signalr hub. 
In my app I have started with Razor pages MVC with default cookie auth. Later I've developed React-based SPA with JWT. I have different controllers which uses auth scheme depending on target front-end. Till that point everything works fine.
App also uses signalr hub, which needs to be accessed by both kind of auth schemes and an anonymous user for few methods. I've created a policy:

```
services.AddAuthorization(options =>
            {
                options.AddPolicy("Combined", new AuthorizationPolicyBuilder()
                     .RequireAuthenticatedUser()
                     .RequireRole(StaticDetails.AdminUser, StaticDetails.DoctorUser)
                     .AddAuthenticationSchemes(CookieAuthenticationDefaults.AuthenticationScheme, JwtBearerDefaults.AuthenticationScheme)
                     .Build());
            });
```
And decorated my hub class like [Authorize(Policy = "Combined"].
With this solution I can access my hub by both kind of authorization, but methods decorated with [AllowAnonymous] cannot be accessed by unauthorized user.
On the other hand if I decorate all methods I want to protect with [Authorize(Policy = "Combined"] (without decorating class) only cookie authenticated user can access hub (anonymous user also works).
I have just updated my app to .NET Core 3.0, but it did not help.

Is that a known issue? I couldn't find any topic covering my problem.

**Comment by blowdart** [comment]
Do both schemes have roles you require in them, or only one?

Also note that .NET Core 3.0 is [EOL](https://dotnet.microsoft.com/platform/support/policy/dotnet-core) and out of support, if you're updating please go to 3.1

**Comment by pbakun** [comment]
Yes, both require roles, but commenting this line doesn't change a thing.

Maybe to clarify I will show how Identity and Authentication is configured.
 
`services.AddIdentity<IdentityUser, IdentityRole>(config =>
            {
                config.Password.RequireNonAlphanumeric = false;
                config.Password.RequireDigit = false;
                config.Password.RequiredLength = 0;
                config.Password.RequiredUniqueChars = 0;
                config.Password.RequireUppercase = false;
                config.Password.RequireLowercase = false;
            })
               .AddRoleManager<RoleManager<IdentityRole>>()
               .AddUserManager<UserManager<IdentityUser>>()
               .AddUserManager<CustomUserManager>()
               .AddDefaultTokenProviders()
               .AddDefaultUI()
               .AddEntityFrameworkStores<RepositoryContext>();

services.ConfigureApplicationCookie(options =>
            {
                options.AccessDeniedPath = "/Identity/Account/AccessDenied";
                options.Cookie.HttpOnly = true;
                options.ExpireTimeSpan = TimeSpan.FromDays(1);
                options.LoginPath = "/Identity/Account/Login";
                options.ReturnUrlParameter = CookieAuthenticationDefaults.ReturnUrlParameter;
                options.SlidingExpiration = true;
                options.ForwardDefault = CookieAuthenticationDefaults.AuthenticationScheme;
            });

            var key = Encoding.ASCII.GetBytes(secretKey);

            services.AddAuthentication(options =>
            {
                options.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                options.DefaultSignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
            })
                .AddCookie(CookieAuthenticationDefaults.AuthenticationScheme, cfg => {
                    cfg.SlidingExpiration = true;
                    cfg.LoginPath = "/Identity/Account/Login";
                })
                .AddJwtBearer(x =>
                {
                    x.RequireHttpsMetadata = false;
                    x.SaveToken = true;
                    x.TokenValidationParameters = new TokenValidationParameters
                    {
                        ValidateIssuerSigningKey = true,
                        IssuerSigningKey = new SymmetricSecurityKey(key),
                        ValidateIssuer = false,
                        ValidateAudience = false
                    };

                    x.Events = new JwtBearerEvents
                    {
                        OnMessageReceived = context =>
                        {
                            var accessToken = context.Request.Query["access_token"];

                            // If the request is for our hub...
                            var path = context.HttpContext.Request.Path;
                            if (!string.IsNullOrEmpty(accessToken))
                            {
                                // Read the token out of the query string
                                context.Token = accessToken;
                            }
                            return Task.CompletedTask;
                        }
                    };
                });
`

**Comment by anurse** [comment]
> And decorated my hub class like [Authorize(Policy = "Combined"].
> With this solution I can access my hub by both kind of authorization, but methods decorated with [AllowAnonymous] cannot be accessed by unauthorized user.

SignalR's authorization model works differently than MVC's, because the client is making a persistent connection to the Hub. In order to connect to a Hub, a client must be able to satisfy the Authorization policy on the Hub. Authorization attributes on Hub methods can only **restrict** access further. You can't use `AllowAnonymous` on a Method to bypass the Hub-level authorization policy.

If you need anonymous users to be able to invoke some methods, there are two recommended approaches here:

* Use a separate Hub for methods that allow anonymous access. This means a separate connection will be established but it helps ensure a good partition of anonymous and authenticated methods.
* Allow anonymous connections to the Hub and put your auth policies on the methods that require authenticated users. This allows a single connection, but is more error-prone since you have to put auth policies on all the methods you want to secure.
